src/allmydata/crypto/ed25519.py

```diff
@@ -0,0 +1,225 @@
+cut-and-pasteability. The base62 encoding is shorter than the base32 form,
+but the minor usability improvement is not worth the documentation and
+specification confusion of using a non-standard encoding. So we stick with
+base32.
+'''
+
+import six
+
+from cryptography.exceptions import (
+    InvalidSignature,
+)
+from cryptography.hazmat.primitives.asymmetric.ed25519 import (
+    Ed25519PrivateKey,
+    Ed25519PublicKey,
+)
+from cryptography.hazmat.primitives.serialization import (
+    Encoding,
+    PrivateFormat,
+    NoEncryption,
+    PublicFormat,
+)
+
+from allmydata.crypto import (
+    remove_prefix,
+    BadSignature,
+)
+from allmydata.util.base32 import (
+    a2b,
+    b2a,
+)
+
+PRIVATE_KEY_PREFIX = 'priv-v0-'
+PUBLIC_KEY_PREFIX = 'pub-v0-'
+
+
+def create_signing_keypair():
+    """
+    Creates a new ed25519 keypair.
+
+    :returns: 2-tuple of (private_key, public_key)
+    """
+    private_key = Ed25519PrivateKey.generate()
+    return private_key, private_key.public_key()
+
+
+def signing_keypair_from_bytes(private_bytes):
+    """
+    Creates a ed25519 keypair from serialized bytes
+
+    :returns: 2-tuple of (private_key, public_key)
+    """
+
+    if not isinstance(private_bytes, six.binary_type):
+        raise ValueError('private_bytes must be bytes')
+    private_key = Ed25519PrivateKey.from_private_bytes(private_bytes)
+    return private_key, private_key.public_key()
+
+
+def bytes_from_signing_key(private_key):
+    """
+    Turn a private signing key into serialized bytes
+    """
+    _validate_private_key(private_key)
+    return private_key.private_bytes(
+        Encoding.Raw,
+        PrivateFormat.Raw,
+        NoEncryption(),
+    )
+
+
+def verifying_key_from_signing_key(private_key):
+    """
+    :returns: the public key associated to the given `private_key`
+    """
+    _validate_private_key(private_key)
+    return private_key.public_key()
+
+
+def sign_data(private_key, data):
+    """
+    Sign the given data using the given private key
+
+    :param private_key: the private part returned from
+        `create_signing_keypair` or from
+        `signing_keypair_from_bytes`
+
+    :param bytes data: the data to sign
+
+    :returns: bytes representing the signature
+    """
+
+    _validate_private_key(private_key)
+    if not isinstance(data, six.binary_type):
+        raise ValueError('data must be bytes')
+    return private_key.sign(data)
+
+
+def string_from_signing_key(private_key):
+    """
+    Encode a private key to a string
+
+    :param private_key: the private part returned from
+        `create_signing_keypair` or from
+        `signing_keypair_from_bytes`
+
+    :returns: string representing this key
+    """
+    _validate_private_key(private_key)
+    return PRIVATE_KEY_PREFIX + b2a(bytes_from_signing_key(private_key))
+
+
+def signing_keypair_from_string(private_key_str):
+    """
+    Load a signing keypair from a string
```

Comment:
bytes?  text?

Reply:
...so, was going to say "should be text" buuuuuut `allmydata.util.base32` appears to be somewhat sloppy about `str` vs `unicode`..? If we go by what `binascii` does in python3, then everything is bytes (i.e. `binascii.b2a_base64(b"something")` returns `bytes`, as does `binascii.a2b_base64(b"c29tZSBieXRlcw==\n")`)

Reply:
Digging further, it seems like there's more than just bytes vs text here as well.  The bytes representation of the key is some opaque binary blob.  The string representation is not _just_ the base32 encoded version of those bytes but has some additional formatting as well - for example, the "`priv-v0-`" prefix.  I wonder if these two forms have names other than "bytes" and "string" (which are quite weak names).

Considering the prefix and the usage - which seems to be "presentation to humans" (eg in `tahoe admin generate-keypair`) - I think text, probably (but indeed we may have to encode/decode bytes explicitly here to achieve that).  And maybe "text" instead of "string" in the API names?  Or possibly this "tag-version-base32" format has some other name we could use (but I don't know what it might be).

Reply:
The only place the `_bytes` things were used was one test and in the `_string` variants, so I just got rid of the different ones and now there's only `_from_string()` functions and they all take and return `bytes` (which I also more-explicitly documented).

Reply:
(Still open to re-naming them if there's a better name to use?)